apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/test/java/gov/cms/bfd/pipeline/rda/grpc/server/FilteredMessageSourceTest.java

```diff
@@ -0,0 +1,62 @@
+package gov.cms.bfd.pipeline.rda.grpc.server;
+
+import static org.junit.Assert.*;
+import static org.mockito.Mockito.*;
+
+import java.util.NoSuchElementException;
+import org.junit.Test;
+
+public class FilteredMessageSourceTest {
+  @Test
+  public void filtering() throws Exception {
+    MessageSource<Integer> source = createSource();
+    MessageSource<Integer> filtered = new FilteredMessageSource<>(source, i -> i != 2);
+    assertTrue(filtered.hasNext());
+    assertEquals(Integer.valueOf(1), filtered.next());
+    assertTrue(filtered.hasNext());
+    assertEquals(Integer.valueOf(3), filtered.next());
+    assertFalse(filtered.hasNext());
+  }
+
+  @Test
+  public void closePassesThrough() throws Exception {
+    MessageSource<Integer> source = createSource();
+    MessageSource<Integer> filtered = new FilteredMessageSource<>(source, i -> i != 2);
+    filtered.close();
+    verify(source).close();
+  }
+
+  @Test
+  public void multipleHashNextOk() throws Exception {
+    MessageSource<Integer> source = createSource();
+    MessageSource<Integer> filtered = new FilteredMessageSource<>(source, i -> i < 10);
+    assertTrue(filtered.hasNext());
+    assertTrue(filtered.hasNext());
+    assertTrue(filtered.hasNext());
+    assertEquals(Integer.valueOf(1), filtered.next());
+
+    assertTrue(filtered.hasNext());
+    assertTrue(filtered.hasNext());
+    assertTrue(filtered.hasNext());
+    verify(source, times(2)).hasNext();
+  }
+
+  @Test(expected = NoSuchElementException.class)
+  public void nextWithoutHasNextFails() throws Exception {
+    MessageSource<Integer> source = createSource();
+    MessageSource<Integer> filtered = new FilteredMessageSource<>(source, i -> i < 10);
+    filtered.next();
```

Comment:
Should this happen?  I think the normal behavior of iterator type APIs is that `next()` will always handle advancing the pointer as well as grabbing the next value, while `hasNext()` simply checks if the cursor is currently pointing to `null`, acting as a simple safeguard with no side effects.

Reply:
The protocol here is that `hasNext()` has to be called before `next()`.  This isn't an `Iterator` so I feel free to deviate from that interface's protocol.